File: internal/worker/translation_failure_test.go
package worker

import (
	"testing"
	"time"

	"github.com/javinizer/javinizer-go/internal/models"
	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
)

func TestIsTranslationFailure(t *testing.T) {
	movie := &models.Movie{ID: "IPX-535"}
	tests := []struct {
		name   string
		result *MovieResult
		want   bool
	}{
		{
			name: "deferred translation failure",
			result: &MovieResult{
				Status: models.JobStatusFailed,
				Error:  "translation stage failed: invalid output",
				Movie:  movie,
			},
			want: true,
		},
		{
			name: "legacy cached translation failure",
			result: &MovieResult{
				Status: models.JobStatusFailed,
				Error:  "translation failed for cached IPX-535: provider unavailable",
				Movie:  movie,
			},
			want: true,
		},
		{
			name: "translation persistence failure is not recoverable",
			result: &MovieResult{
				Status: models.JobStatusFailed,
				Error:  "translation checkpoint persistence failed",
				Movie:  movie,
			},
		},
		{
			name: "scrape failure",
			result: &MovieResult{
				Status: models.JobStatusFailed,
				Error:  "scrape failed: no metadata",
				Movie:  movie,
			},
		},
		{
			name: "missing retained movie",
			result: &MovieResult{
				Status: models.JobStatusFailed,
				Error:  "translation stage failed: invalid output",
			},
		},
	}

	for _, tc := range tests {
		t.Run(tc.name, func(t *testing.T) {
			assert.Equal(t, tc.want, IsTranslationFailure(tc.result))
		})
	}
}

func TestJobEditorResolveTranslationFailure(t *testing.T) {
	store := NewJobStore(nil, nil, nil, "", nil, nil)
	job := store.CreateJobBatch([]string{"/media/IPX-535.mp4"})
	resultID := "translation-failure"
	warning := "old warning"
	job.results.UpdateFileResult("/media/IPX-535.mp4", &MovieResult{
		ResultID: resultID,
		FileMatchInfo: models.FileMatchInfo{
			Path:    "/media/IPX-535.mp4",
			MovieID: "IPX-535",
		},
		Status:             models.JobStatusFailed,
		Error:              "translation stage failed: invalid title",
		Movie:              &models.Movie{ID: "IPX-535", Title: "번역 제목"},
		OrchestrationState: models.OrchestrationState{TranslationWarning: &warning},
		StartedAt:          time.Now(),
	})

	controlled, ok := store.GetBatchJob(job.GetID())
	require.True(t, ok)
	updated, err := controlled.ResolveTranslationFailure(resultID)

	require.NoError(t, err)
	require.NotNil(t, updated)
	assert.Equal(t, models.JobStatusCompleted, updated.Status)
	assert.Empty(t, updated.Error)
	assert.Nil(t, updated.TranslationWarning)
	status := controlled.GetStatus()
	assert.Equal(t, 1, status.Completed)
	assert.Zero(t, status.Failed)

	_, err = controlled.ResolveTranslationFailure(resultID)
	assert.ErrorContains(t, err, "not a recoverable translation failure")
}

File: web/frontend/src/lib/utils/translation-failure.test.ts
import { describe, expect, it } from 'vitest';
import type { FileResult } from '$lib/api/types';
import {
	canRetranslateBatchJob,
	countTranslationFailures,
	isTranslationFailure,
} from './translation-failure';

function result(overrides: Partial<FileResult> = {}): FileResult {
	return {
		result_id: 'result-1',
		file_path: '/media/IPX-535.mp4',
		movie_id: 'IPX-535',
		status: 'failed',
		movie: { id: 'IPX-535', title: '日本語題名' },
		started_at: new Date(0).toISOString(),
		is_multi_part: false,
		part_number: 0,
		part_suffix: '',
		...overrides,
	};
}

describe('isTranslationFailure', () => {
	it('recognizes recoverable deferred translation failures', () => {
		expect(
			isTranslationFailure(
				result({ error: 'translation stage failed: invalid title output' }),
			),
		).toBe(true);
	});

	it('does not classify translation checkpoint persistence failures', () => {
		expect(
			isTranslationFailure(
				result({ error: 'translation checkpoint persistence failed' }),
			),
		).toBe(false);
	});

	it('requires retained movie metadata', () => {
		expect(
			isTranslationFailure(
				result({
					error: 'translation stage failed: provider unavailable',
					movie: undefined,
				}),
			),
		).toBe(false);
	});
});

describe('countTranslationFailures', () => {
	it('counts only recoverable translation failures', () => {
		expect(
			countTranslationFailures({
				results: {
					'/media/IPX-535.mp4': result({
						error: 'translation stage failed: invalid title output',
					}),
					'/media/ABC-123.mp4': result({
						result_id: 'result-2',
						file_path: '/media/ABC-123.mp4',
						movie_id: 'ABC-123',
						error: 'scrape failed: no metadata',
					}),
				},
			}),
		).toBe(1);
	});
});

describe('canRetranslateBatchJob', () => {
	it('allows completed jobs with a recoverable translation failure', () => {
		expect(
			canRetranslateBatchJob({
				status: 'completed',
				completed: 0,
				results: {
					'/media/IPX-535.mp4': result({
						error: 'translation stage failed: provider unavailable',
					}),
				},
			}),
		).toBe(true);
	});

	it('hides retranslation after organization', () => {
		expect(
			canRetranslateBatchJob({
				status: 'organized',
				completed: 1,
				results: {
					'/media/IPX-535.mp4': result({
						status: 'completed',
						error: undefined,
					}),
				},
			}),
		).toBe(false);
	});
});
